### PR TITLE
Upgrade to webpack(er) 4.

### PR DESCRIPTION
**Pull request by jfly**
Following the instructions on
https://github.com/rails/webpacker/blob/master/docs/v4-upgrade.md.

It looks like Webpacker 4 generates some helpful warnings we should probably pay attention to:

```
WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets:
  media/1x1/do-e31ff1fd.svg (383 KiB)
  media/4x3/do-06e1cfe3.svg (385 KiB)
  js/auto_numeric-801c07f4c071c37e03f2.js (283 KiB)
  js/edit_events-1206a91cbd0bf34f380e.js (289 KiB)
  js/edit_schedule-e9e3ff1b2a066f669863.js (461 KiB)
  js/markdown_editor-125f73a759da18f526b8.js (287 KiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  auto_numeric (283 KiB)
      js/auto_numeric-801c07f4c071c37e03f2.js
  edit_events (289 KiB)
      js/edit_events-1206a91cbd0bf34f380e.js
  edit_schedule (461 KiB)
      js/edit_schedule-e9e3ff1b2a066f669863.js
  markdown_editor (296 KiB)
      css/markdown_editor-22647a5f.css
      js/markdown_editor-125f73a759da18f526b8.js

WARNING in webpack performance recommendations:
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/
Child mini-css-extract-plugin node_modules/css-loader/dist/cjs.js??ref--6-1!node_modules/postcss-loader/src/index.js??ref--6-2!node_modules/flag-icon-css/css/flag-icon.css:
    Entrypoint mini-css-extract-plugin = *
       517 modules
Child mini-css-extract-plugin node_modules/css-loader/dist/cjs.js??ref--6-1!node_modules/postcss-loader/src/index.js??ref--6-2!node_modules/simplemde/debug/simplemde.css:
    Entrypoint mini-css-extract-plugin = *
       2 modules
Child mini-css-extract-plugin node_modules/css-loader/dist/cjs.js??ref--7-1!node_modules/postcss-loader/src/index.js??ref--7-2!node_modules/sass-loader/lib/loader.js??ref--7-3!app/javascript/markdown-editor/style.scss:
    Entrypoint mini-css-extract-plugin = *
    [0] ./node_modules/css-loader/dist/cjs.js??ref--7-1!./node_modules/postcss-loader/src??ref--7-2!./node_modules/sass-loader/lib/loader.js??ref--7-3!./app/javascript/markdown-editor/style.scss 631 bytes {0} [built]
        + 1 hidden module
```

(@viroulep in case I broke some of the work you did to split up our
javascript into smaller modules)

### CODE REVIEW

**Comment by jfly** [comment]
I've deployed this to staging and it seems to work!

**Comment by viroulep** [comment]
I remember I added a script target in the package.json, it generates an image showing the size of each package relative to the whole pack size.
Could you please check that it still works and that it's consistent with what it was before?

**Comment by jfly** [comment]
Ah, good call!

## Before upgrading to webpack(er) 4

image

## After upgrading to webpack(er) 4

Yikes! Our bundle sizes did go up a bit. 1.33 MB -> 1.47 MB total, and particularly noticeable in show_schedule (37.05 KB -> 82.19 KB). Looking closely, it looks like we're now pulling in all of lodash instead of some of it:

image

## After adding `lodash` plugin to babel.config.js

I added "lodash" to the list of babel plugins in babel.config.js, and things look better now! The total seems to have actually dropped to 1.3 MB.

image

**Comment by viroulep** [comment]
Awesome :)
The lodash plugin is indeed something that automatically only includes what is needed from lodash in each of our packs.

File: WcaOnRails/package.json [hunk 1]
```diff
@@ -52,6 +52,6 @@
     "webpack-dev-server": "^3.2.1"
   },
   "scripts": {
-    "webpack:analyze": "mkdir -p public/packs && NODE_ENV=production node_modules/.bin/webpack --config config/webpack/production.js --profile --json > public/packs/stats.json && node_modules/.bin/webpack-bundle-analyzer public/packs/stats.json"
+    "webpack:analyze": "mkdir -p public/packs && RACK_ENV=test NODE_ENV=production webpack --config config/webpack/production.js --profile --json > public/packs/stats.json && webpack-bundle-analyzer public/packs/stats.json"
```
Comment (viroulep) [inline WcaOnRails/package.json]: Hum wouldn't that only work if `node_modules/bin` is in your `PATH`?
(or if webpack was installed "globally" on the computer)

Comment (jfly) [reply WcaOnRails/package.json]: Ah, npm/yarn is actually nice enough to set up your `PATH` for you when you run a script! See https://docs.npmjs.com/misc/scripts#path. This works for me and I do *not* have `node_modules/bin` in my `PATH`, nor do I have webpack installed globally =)

**Comment by jfly** [comment]
> Our bin/webpack-dev-server has been broken for a while now, does this make it work again?

Oh WOW yes it does and holy COW is it fast! Changing a jsx file results in an update on the page nearly instantly!

File: WcaOnRails/package.json (hunk 1 above)
Comment (viroulep) [reply WcaOnRails/package.json]: Awesome!

**Pull request merged**